tests/blast/test_blast_output.h

```diff
@@ -0,0 +1,814 @@
+// Copyright (c) 2006-2015, Knut Reinert, FU Berlin
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ==========================================================================
+// Author: Hannes Hauswedell <hannes.hauswedell@fu-berlin.de>
+// ==========================================================================
+// Tests for the blast module
+// ==========================================================================
+
+#ifndef SEQAN_TESTS_TEST_BLAST_OUTPUT_H_
+#define SEQAN_TESTS_TEST_BLAST_OUTPUT_H_
+
+#include <seqan/basic.h>
+#include <seqan/sequence.h>
+
+#include <seqan/blast.h>
+
+using namespace seqan;
+
+template <typename TFile, typename TScore, typename TString, typename TRecords, BlastProgram p, BlastTabularSpec h>
+inline void
+test_blast_write_do(TFile & file,
+                    BlastIOContext<TScore, TString, p, h> & context,
+                    TRecords const & records,
+                    int const format,
+                    int const,
+                    BlastReport const & /**/)
+{
+    if (format < 3)
+    {
+        writeHeader(file, context, BlastReport());
+        for (auto const & r : records)
+            writeRecord(file, context, r, BlastReport());
+        writeFooter(file, context, BlastReport());
+    } else
+    {
+        BlastReportOut<BlastIOContext<TScore, TString, p, h>> out(context, file, BlastReport());
```

Comment:
Not really. It has to be templatized somewhere. Only templatizing the context at least moves the variability to one place.
Since downstream functions don't work on the FormattedFile but on the context and the iterator, obviously we can't move sepcializations from the context to the FormattedFile.